calendar/src/main/java/com/toggl/calendar/calendarday/ui/views/CalendarItemDrawingDelegate.kt

```diff
@@ -1,21 +1,337 @@
+    var currentHourHeight: Float = 0f
+        set(value) {
+            minHourHeight = value / 4
+            field = value
+        }
 
     fun onDraw(canvas: Canvas, calendarBounds: RectF, calendarItem: CalendarItem) {
-        calendarItem.run {
-            drawingRect.set(calendarBounds.left, calendarBounds.top, calendarBounds.right, 300f)
-            canvas.drawRect(drawingRect, itemPaint)
+        calculateItemRect(calendarItem, calendarBounds)
+        when {
+            calendarItem is CalendarItem.TimeEntry && calendarItem.duration() != null ->
+                drawStoppedTimeEntryItem(canvas, calendarItem)
+            calendarItem is CalendarItem.TimeEntry && calendarItem.duration() == null ->
+                drawRunningTimeEntryItem(canvas, calendarItem)
+            calendarItem is CalendarItem.CalendarEvent -> drawCalendarEventItem(canvas, calendarItem)
         }
+        drawCalendarItemText(canvas, calendarItem)
+    }
+
+    private fun drawCalendarItemText(canvas: Canvas, calendarItem: CalendarItem) {
+        val textLeftPadding =
+            if (calendarItem is CalendarItem.CalendarEvent) calendarIconSize - calendarIconRightInsetMargin else 0f
+        val eventHeight = drawingRect.height()
+        val eventWidth = drawingRect.width() - textLeftPadding
+        val fontSize =
+            if (eventHeight <= shortCalendarItemHeight) shortCalendarItemFontSize else regularCalendarItemFontSize
+        var textVerticalPadding =
+            if (eventHeight <= shortCalendarItemHeight) shortCalendarItemVerticalPadding else regularCalendarItemVerticalPadding
+        textVerticalPadding = kotlin.math.min(eventHeight - fontSize, textVerticalPadding)
+        val textHorizontalPadding =
+            if (eventHeight <= shortCalendarItemHeight) shortCalendarItemHorizontalPadding else regularCalendarItemHorizontalPadding
+
+        val textWidth = eventWidth - textHorizontalPadding * 2
+        if (textWidth <= 0) return
+
+        val isRunning = calendarItem.duration() == null
+        val eventTextLayout = getCalendarItemTextLayout(calendarItem, textWidth, fontSize, isRunning)
+        val totalLineHeight = calculateLineHeight(eventHeight, eventTextLayout).toFloat()
+
+        canvas.save()
+        canvas.translate(
+            drawingRect.left + textHorizontalPadding + textLeftPadding,
+            drawingRect.top + textVerticalPadding
+        )
+        canvas.clipRect(0f, 0f, eventWidth - textHorizontalPadding, totalLineHeight)
+        eventTextLayout.draw(canvas)
+        canvas.restore()
+    }
+
+    private fun calculateLineHeight(eventHeight: Float, eventTextLayout: StaticLayout): Int {
+        var totalLineHeight = 0
+        for (i in 0 until eventTextLayout.lineCount) {
+            val lineBottom = eventTextLayout.getLineBottom(i)
+            if (lineBottom <= eventHeight) {
+                totalLineHeight = lineBottom
+            }
+        }
+
+        return totalLineHeight
+    }
```

Comment:
```suggestion
    private fun calculateLineHeight(eventHeight: Float, eventTextLayout: StaticLayout): Int = 
        (0 until eventTextLayout.lineCount)
            .map(eventTextLayout::getLineBottom)
            .filter { it <= eventHeight }
            .max() ?: 0
```

This should be a var-less equivalent right?

Reply:
Yeah, we can make it shorter with `maxBy` as well. I'll add this.